lib/Dialect/LLHD/Transforms/DesequentializationPass.cpp

```diff
@@ -0,0 +1,721 @@
+  if (failed(isSupportedSequentialProcess(procOp, trAnalysis)))
+    return;
+
+  OpBuilder builder(procOp);
+  WalkResult result = procOp.walk([&](llhd::DrvOp op) {
+    LLVM_DEBUG({ llvm::dbgs() << "\n  Lowering Drive Operation\n"; });
+
+    if (!op.getEnable()) {
+      LLVM_DEBUG({ llvm::dbgs() << "  - No enable condition -> skip\n"; });
+      return WalkResult::advance();
+    }
+
+    Location loc = op.getLoc();
+    builder.setInsertionPoint(op);
+    int presentTR = trAnalysis.getBlockTR(op.getOperation()->getBlock());
+
+    auto sampledInPast = [&](Value value) -> bool {
+      if (isa<BlockArgument>(value))
+        return false;
+
+      if (!procOp->isAncestor(value.getDefiningOp()))
+        return false;
+
+      return trAnalysis.getBlockTR(value.getDefiningOp()->getBlock()) !=
+             presentTR;
+    };
+
+    LLVM_DEBUG({ llvm::dbgs() << "  - Analyzing enable condition...\n"; });
+
+    SmallVector<Trigger> triggers;
+    auto sampledFromSameSignal = [](Value val1, Value val2) -> bool {
+      if (auto prb1 = val1.getDefiningOp<llhd::PrbOp>())
+        if (auto prb2 = val2.getDefiningOp<llhd::PrbOp>())
+          return prb1.getSignal() == prb2.getSignal();
+
+      // TODO: consider signals not represented by hw.inout (and thus don't have
+      // an llhd.prb op to look at)
+      return false;
+    };
+
+    if (failed(DnfAnalyzer(op.getEnable(), sampledInPast)
+                   .computeTriggers(builder, loc, sampledFromSameSignal,
+                                    triggers))) {
+      LLVM_DEBUG({
+        llvm::dbgs() << "  Unable to compute trigger list for drive condition, "
+                        "skipping...\n";
+      });
+      return WalkResult::interrupt();
+    }
+
+    LLVM_DEBUG({
+      if (triggers.empty())
+        llvm::dbgs() << "  - no triggers found!\n";
+    });
+
+    LLVM_DEBUG({
+      for (auto trigger : triggers) {
+        llvm::dbgs() << "  - Trigger\n";
+        for (auto [clk, kind] : llvm::zip(trigger.clocks, trigger.kinds))
+          llvm::dbgs() << "    - " << kind << " "
+                       << "clock: " << clk << "\n";
+
+        if (trigger.enable)
+          llvm::dbgs() << "      with enable: " << trigger.enable << "\n";
+      }
+    });
+
+    // TODO: add support
+    if (triggers.size() > 2 || triggers.empty())
+      return WalkResult::interrupt();
+
+    // TODO: add support
+    if (triggers[0].clocks.size() != 1 || triggers[0].clocks.size() != 1)
+      return WalkResult::interrupt();
+
+    // TODO: add support
+    if (triggers[0].kinds[0] == Trigger::Kind::Edge)
+      return WalkResult::interrupt();
+
+    Value clock = builder.create<seq::ToClockOp>(loc, triggers[0].clocks[0]);
+    Value reset, resetValue;
+
+    if (triggers[0].kinds[0] == Trigger::Kind::NegEdge)
+      clock = builder.create<seq::ClockInverterOp>(loc, clock);
+
+    if (triggers[0].enable)
+      clock = builder.create<seq::ClockGateOp>(loc, clock, triggers[0].enable);
```

Comment:
Nice idea to use clock inverters and clock gates to materialize the different trigger conditions :+1: